core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -257,8 +256,8 @@ private CompletionStage<CacheStatusResponse> sendJoinRequest(String cacheName, C
             log.debugf("Join request received CacheNotFoundResponse for cache %s, retrying", cacheName);
          } else {
             log.debugf(t, "Join request failed for cache %s", cacheName);
-            if (t instanceof TimeoutException) {
-               throw (TimeoutException) t;
+            if (t instanceof RuntimeException && !(t instanceof CacheJoinException)) {
+               throw (RuntimeException) t;
```

Comment:
Could you instead add a new if block checking whether the `t.getCause()` is an instance of `CacheJoinException` and re-throw it as a `CacheException`? This way we keep the timeout handling, too.

Reply:
Oh yeah, I wasn't checking the getCause()! Good spot!

I shuffled it around a bit to make it a bit easier to understand.

Is this what you had in mind? This would mean anything that's not a CacheJoinException would be wrapped in a CacheException (including the TimeoutException).